From 9c7571646a01eedb85350dfce12b499a0267ab2b Mon Sep 17 00:00:00 2001
From: Hari Chandrakanthan <quic_haric@quicinc.com>
Date: Thu, 3 Feb 2022 14:01:57 +0530
Subject: [PATCH] mac80211 : fix bw change to 40Mhz during channel switch

When AP reduces its channel bandwidth to 40Mhz, the associated
sta reduces the channel bandwidth to 20Mhz.

From spec 802.11 ac, section 8.4.2.165 :
The Wide Bandwidth Channel Switch subelement is present under the following conditions:
1.Channel switching to a BSS operating channel width of 40 MHz or wider
2.Extended channel switching to a BSS operating channel width of 80 MHz or wider

So when wide bandwidth channel switch subelement is present,
the default bandwidth is chosen as 40Mhz.

Signed-off-by: Hari Chandrakanthan <quic_haric@quicinc.com>
---
 net/mac80211/spectmgmt.c | 4 ++++
 1 file changed, 4 insertions(+)

--- a/net/mac80211/spectmgmt.c
+++ b/net/mac80211/spectmgmt.c
@@ -136,6 +136,10 @@ int ieee80211_parse_ch_switch_ie(struct
 	}
 
 	if (wide_bw_chansw_ie) {
+		csa_ie->chandef.width = NL80211_CHAN_WIDTH_40;
+		csa_ie->chandef.center_freq1 =
+			ieee80211_channel_to_frequency(wide_bw_chansw_ie->new_center_freq_seg0,
+						       new_chan->band);
 		new_seg1 = wide_bw_chansw_ie->new_center_freq_seg1;
 		vht_oper.chan_width = wide_bw_chansw_ie->new_channel_width;
 		vht_oper.center_freq_seg0_idx = wide_bw_chansw_ie->new_center_freq_seg0;
